cmd/sync-rover-groups/main.go

```diff
@@ -0,0 +1,164 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	ldapv3 "github.com/go-ldap/ldap/v3"
+	"github.com/sirupsen/logrus"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/test-infra/prow/flagutil"
+	"k8s.io/test-infra/prow/logrusutil"
+	"sigs.k8s.io/yaml"
+
+	templatev1 "github.com/openshift/api/template/v1"
+	userv1 "github.com/openshift/api/user/v1"
+
+	"github.com/openshift/ci-tools/pkg/api"
+)
+
+type options struct {
+	logLevel         string
+	manifestDirRaw   flagutil.Strings
+	manifestDirs     sets.String
+	ldapServer       string
+	validateSubjects bool
+	groupsFile       string
+}
+
+func parseOptions() *options {
+	opts := &options{}
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+	fs.StringVar(&opts.logLevel, "log-level", "info", fmt.Sprintf("Log level is one of %v.", logrus.AllLevels))
+	fs.Var(&opts.manifestDirRaw, "manifest-dir", "directory containing Kubernetes manifests. Can be specified multiple times.")
```

Comment:
Iterating over manifests to figure out the groups that should exist is smart, I like how it avoids a necessity to maintain a separate config file or something similar. Two concerns:

- Do we expect the need to support groups that are supposed to have different names in Rover and in Kubernetes (are the two naming rules fully compatible)?
- Do we expect the need to support certain groups to only exist on certain clusters? I imagine this is a legit request for e.g. the Hive cluster which will need to provide access to groups to their pools, but these groups are not necessary on normal build farm clsuters (also, app.ci is probably unique too).

In my head, I was expecting a config file to drive these use cases, but maybe they are not possible.

---

As an entirely wild thought: maybe this should be a controller? Imagine the following:
- We would have a CRD called `RoverGroup`. Creating such resource in a cluster would represent the intent of creating a `Group` of the same name, and maintaining its members to be equivalent to the Rover Group
- We would have a controller running in the PSI cluster that would watch `RoverGroup` events on our clusters and would maintain a matching `Group` in the same cluster (by reaching to LDAP, getting the members list, etc). If the `RoverGroup` would be deleted, then it would delete the matching `Group` 
- The controller would also periodically poll LDAP for members of the groups it maintains on the cluster and updated the menbers in clusters accordingly. 

The above ^^^ is just a wild thought. Please do not immediately jump and implement it unless you really like it. It also does not solve the groups for the GitHub<->RH pairing.

Reply:
The 2 cases are valid. The config file is a supplement of what cannot be covered by the auto-detection. What about this?

```yaml
some-rover-group:
  clusters: # default to all clusters
  - app.ci
  - build01
  name: # default to the one at rover 
```

About the controller: Functionality wise, it is the same as the current implementation. It is more efficient in the sense that it is realtime: Do we need realtime? The groups should not change that often. My concern is the maintenance of such a tool: we do not GitOps on psi and secrets there are out of control of ci-secret-bootstrap. The effort on implementing a controller watching multiple clusters is not trivial either.

I am not worrying about the groups for the GitHub<->RH pairing: The current implementation (https://github.com/openshift/ci-tools/tree/master/cmd/ldap-users-from-github-owners-files) can easily included in the sync-rover-groups (no matter to it is a simple binary or a controller).

Reply:
Yeah, it looks like advanced behavior could be added with a config file like that in the future when it is needed. No need to add that to this PR, it's already large.

Reply:
I will do the config file with a follow up PR. This one is big enough.